components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -93,98 +118,330 @@ std::optional<M> GetUintFromData(const std::vector<uint8_t>& input,
+                         consumed);
 }
 
 // GetBoolFromData extracts a 32-byte wide boolean value from the
-// calldata at the specified offset.
-//
-// The parsed bool value is serialized as "true" or "false" strings.
-std::optional<std::string> GetBoolFromData(const std::vector<uint8_t>& input,
-                                           size_t offset) {
-  auto value = GetUintFromData<uint8_t>(input, offset);
-  if (!value) {
+// calldata segment.
+std::optional<DecoderResult<base::Value>> GetBoolFromData(
+    const ByteArray& input) {
+  auto result = GetUintFromData<uint8_t>(input);
+  if (!result) {
     return std::nullopt;
   }
 
+  auto [value, remaining, consumed] = *result;
+
   if (value == static_cast<uint8_t>(0)) {
-    return "false";
+    return std::make_tuple(base::Value(false), remaining, consumed);
   } else if (value == static_cast<uint8_t>(1)) {
-    return "true";
+    return std::make_tuple(base::Value(true), remaining, consumed);
   }
 
   return std::nullopt;
 }
 
-// GetBytesHexFromData extracts a bytes value from the calldata at the
-// specified offset using head-tail encoding mechanism. bytes are packed
-// tightly in chunks of 32 bytes, with the first 32 bytes encoding the length,
-// followed by the actual content.
+// GetBytesHexFromData extracts a bytes value from the calldata segment using
+// head-tail encoding mechanism. bytes are packed in chunks of 32 bytes, with
+// the first 32 bytes encoding the length, followed by the actual content.
 //
-// The parsed bytearray is serialized as a hex string prefixed by "0x".
-std::optional<std::string> GetBytesHexFromData(
-    const std::vector<uint8_t>& input,
-    size_t offset) {
-  auto pointer = GetUintFromData<size_t>(input, offset);
-  if (!pointer) {
+// The first argument indicates the type of the bytes value to be extracted.
+// If the Type::m property is set, it indicates a fixed-size bytes<M> type,
+// where 0 < M <= 32, otherwise it indicates a dynamic bytes type.
+//
+// The result is serialized as a hex string prefixed by "0x".
+std::optional<DecoderResult<base::Value>> GetBytesHexFromData(
+    const eth_abi::Type& type,
+    const ByteArray& input) {
+  size_t size = type.m.value_or(0);
+  if (size > kWordSize) {
     return std::nullopt;
   }
 
-  auto bytes_len = GetUintFromData<size_t>(input, *pointer);
-  if (!bytes_len) {
-    return std::nullopt;
+  ByteArray remaining(input.begin(), input.end());
+  size_t parts_count = 1;
+  size_t consumed = 0;
+
+  // dynamic bytes
+  if (!type.m.has_value()) {
+    // Extract the length of the bytes since it is a dynamic type
+    auto length_result = GetUintFromData<size_t>(remaining);
+    if (!length_result) {
+      return std::nullopt;
+    }
+
+    size = std::get<0>(*length_result);
+
+    remaining = std::get<1>(*length_result);
+    consumed += std::get<2>(*length_result);
+    parts_count = std::ceil(static_cast<double>(size) / kWordSize);
```

Comment:
can we avoid casting to double by `parts_count = (size + kWordSize - 1) / kWordSize;` ?

Reply:
assumimg `size` is adequate

Reply:
The two statements are equivalent. I verified for all 32-bit unsigned integers using a Python script.

4de5c1a824504859c32e92d6dc27c7d8ec708f9f